File: nodejs/test/e2e/compaction.e2e.test.ts
import { describe, expect, it } from "vitest";
import { approveAll, type CopilotSession, type SessionEvent } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const compactionTimeoutMs = 60_000;

function getNextSessionEvent<TEventType extends SessionEvent["type"]>(
    session: CopilotSession,
    eventType: TEventType,
    description: string,
    predicate: (event: Extract<SessionEvent, { type: TEventType }>) => boolean = () => true
): Promise<Extract<SessionEvent, { type: TEventType }>> {
    return new Promise((resolve, reject) => {
        let unsubscribe: () => void = () => {};
        const timeout = setTimeout(() => {
            unsubscribe();
            reject(new Error(`Timed out waiting for ${description}`));
        }, compactionTimeoutMs);

        unsubscribe = session.on((event) => {
            if (event.type === eventType) {
                const typedEvent = event as Extract<SessionEvent, { type: TEventType }>;
                if (predicate(typedEvent)) {
                    clearTimeout(timeout);
                    unsubscribe();
                    resolve(typedEvent);
                }
            } else if (event.type === "session.error") {
                clearTimeout(timeout);
                unsubscribe();
                reject(new Error(`${event.data.message}\n${event.data.stack}`));
            }
        });
    });
}

describe("Compaction", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("should trigger compaction with low threshold and emit events", async () => {
        // Create session with very low compaction thresholds to trigger compaction quickly
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            infiniteSessions: {
                enabled: true,
                // Trigger background compaction at 0.5% context usage (~1000 tokens)
                backgroundCompactionThreshold: 0.005,
                // Block at 1% to ensure compaction runs
                bufferExhaustionThreshold: 0.01,
            },
        });

        // The first prompt leaves the session below the compaction processor's minimum
        // message count. The second prompt is therefore the first deterministic point
        // at which low thresholds can trigger compaction. Register event waiters before
        // any prompts are sent so we never miss the events.
        const compactionStartedP = getNextSessionEvent(
            session,
            "session.compaction_start",
            "session.compaction_start"
        );
        // Wait specifically for a *successful* compaction_complete so that any transient
        // failed compaction event the daemon may emit before a successful retry is ignored
        // (mirrors the dotnet/rust references).
        const compactionCompletedP = getNextSessionEvent(
            session,
            "session.compaction_complete",
            "successful session.compaction_complete",
            (event) => event.data.success
        );

        await session.sendAndWait({
            prompt: "Tell me a story about a dragon. Be detailed.",
        });
        await session.sendAndWait({
            prompt: "Continue the story with more details about the dragon's castle.",
        });

        const [startEvent, completeEvent] = await Promise.all([
            compactionStartedP,
            compactionCompletedP,
        ]);

        expect(startEvent.data.conversationTokens ?? 0).toBeGreaterThan(0);
        expect(completeEvent.data.success).toBe(true);
        expect(completeEvent.data.compactionTokensUsed).toBeDefined();
        expect(completeEvent.data.compactionTokensUsed?.inputTokens ?? 0).toBeGreaterThan(0);
        const summary = (completeEvent.data.summaryContent ?? "").toLowerCase();
        expect(summary).toContain("<overview>");
        expect(summary).toContain("<history>");
        expect(summary).toContain("<checkpoint_title>");

        await session.sendAndWait({
            prompt: "Now describe the dragon's treasure in great detail.",
        });

        // Verify the session still works after compaction
        const answer = await session.sendAndWait({ prompt: "What was the story about?" });
        const content = (answer?.data.content ?? "").toLowerCase();
        // Should remember it was about a dragon (context preserved via summary)
        expect(content).toContain("kaedrith");
        expect(content).toContain("dragon");
    }, 120000);

    it("should not emit compaction events when infinite sessions disabled", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            infiniteSessions: {
                enabled: false,
            },
        });

        const compactionEvents: SessionEvent[] = [];
        session.on((event) => {
            if (
                event.type === "session.compaction_start" ||
                event.type === "session.compaction_complete"
            ) {
                compactionEvents.push(event);
            }
        });

        await session.sendAndWait({ prompt: "What is 2+2?" });

        // Should not have any compaction events when disabled
        expect(compactionEvents.length).toBe(0);
    });

    it("should return empty handoff summary for fresh session", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const result = await session.rpc.history.summarizeForHandoff();
            expect(result.summary).toBe("");
        } finally {
            await session.disconnect();
        }
    });

    it("should summarize for handoff after non-ephemeral log event", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            await session.log("handoff summary log coverage");
            const result = await session.rpc.history.summarizeForHandoff();
            expect(typeof result.summary).toBe("string");
        } finally {
            await session.disconnect();
        }
    });

    it("should report no-op when cancelling compaction without in-flight work", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        try {
            const backgroundResult = await session.rpc.history.cancelBackgroundCompaction();
            const manualResult = await session.rpc.history.abortManualCompaction();

            expect(backgroundResult.cancelled).toBe(false);
            expect(manualResult.aborted).toBe(false);
        } finally {
            await session.disconnect();
        }
    });
});

File: go/internal/e2e/compaction_e2e_test.go
package e2e

import (
	"errors"
	"strings"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestCompactionE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should trigger compaction with low threshold and emit events", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		enabled := true
		backgroundThreshold := 0.005 // 0.5%
		bufferThreshold := 0.01      // 1%

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			InfiniteSessions: &copilot.InfiniteSessionConfig{
				Enabled:                       &enabled,
				BackgroundCompactionThreshold: &backgroundThreshold,
				BufferExhaustionThreshold:     &bufferThreshold,
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// The first prompt leaves the session below the compaction processor's minimum
		// message count. The second prompt is therefore the first deterministic point
		// at which low thresholds can trigger compaction. Subscribe before any prompts
		// are sent so we never miss the events. The complete-event subscription filters
		// for Success==true so any transient failed compaction event the daemon may emit
		// before a successful retry is ignored (mirrors the dotnet/rust references).
		startCh := make(chan copilot.SessionEvent, 1)
		completeCh := make(chan copilot.SessionEvent, 1)
		errCh := make(chan error, 1)
		unsubscribe := session.On(func(event copilot.SessionEvent) {
			switch d := event.Data.(type) {
			case *copilot.SessionCompactionStartData:
				select {
				case startCh <- event:
				default:
				}
			case *copilot.SessionCompactionCompleteData:
				if !d.Success {
					return
				}
				select {
				case completeCh <- event:
				default:
				}
			case *copilot.SessionErrorData:
				msg := d.Message
				if msg == "" {
					msg = "session error"
				}
				select {
				case errCh <- errors.New(msg):
				default:
				}
			}
		})
		defer unsubscribe()

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Tell me a story about a dragon. Be detailed."})
		if err != nil {
			t.Fatalf("Failed to send first message: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Continue the story with more details about the dragon's castle."})
		if err != nil {
			t.Fatalf("Failed to send second message: %v", err)
		}

		const compactionTimeout = 60 * time.Second

		var startEvent copilot.SessionEvent
		select {
		case startEvent = <-startCh:
		case err := <-errCh:
			t.Fatalf("Session error waiting for session.compaction_start event: %v", err)
		case <-time.After(compactionTimeout):
			t.Fatalf("Timed out waiting for session.compaction_start event")
		}

		var completeEvent copilot.SessionEvent
		select {
		case completeEvent = <-completeCh:
		case err := <-errCh:
			t.Fatalf("Session error waiting for session.compaction_complete event: %v", err)
		case <-time.After(compactionTimeout):
			t.Fatalf("Timed out waiting for session.compaction_complete event")
		}

		startData, ok := startEvent.Data.(*copilot.SessionCompactionStartData)
		if !ok {
			t.Fatalf("Expected SessionCompactionStartData, got %T", startEvent.Data)
		}
		if startData.ConversationTokens == nil || *startData.ConversationTokens <= 0 {
			t.Errorf("Expected compaction to report conversation tokens at start, got %v", startData.ConversationTokens)
		}

		completeData, ok := completeEvent.Data.(*copilot.SessionCompactionCompleteData)
		if !ok {
			t.Fatalf("Expected SessionCompactionCompleteData, got %T", completeEvent.Data)
		}
		if !completeData.Success {
			t.Errorf("Expected compaction to succeed, error=%v", completeData.Error)
		}
		if completeData.CompactionTokensUsed == nil {
			t.Errorf("Expected compaction tokens-used data")
		} else if completeData.CompactionTokensUsed.InputTokens == nil || *completeData.CompactionTokensUsed.InputTokens <= 0 {
			t.Errorf("Expected compaction call to consume input tokens, got %v", completeData.CompactionTokensUsed.InputTokens)
		}
		summary := ""
		if completeData.SummaryContent != nil {
			summary = *completeData.SummaryContent
		}
		summary = strings.ToLower(summary)
		if !strings.Contains(summary, "<overview>") {
			t.Errorf("Expected summary to contain <overview>, got: %q", summary)
		}
		if !strings.Contains(summary, "<history>") {
			t.Errorf("Expected summary to contain <history>, got: %q", summary)
		}
		if !strings.Contains(summary, "<checkpoint_title>") {
			t.Errorf("Expected summary to contain <checkpoint_title>, got: %q", summary)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Now describe the dragon's treasure in great detail."})
		if err != nil {
			t.Fatalf("Failed to send third message: %v", err)
		}

		// Verify session still works after compaction
		answer, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What was the story about?"})
		if err != nil {
			t.Fatalf("Failed to send verification message: %v", err)
		}
		ad, ok := answer.Data.(*copilot.AssistantMessageData)
		if !ok {
			t.Fatalf("Expected assistant message data, got %T", answer.Data)
		}
		content := strings.ToLower(ad.Content)
		// Should remember it was about a dragon (context preserved via summary)
		if !strings.Contains(content, "kaedrith") {
			t.Errorf("Expected answer to mention 'Kaedrith', got: %q", ad.Content)
		}
		if !strings.Contains(content, "dragon") {
			t.Errorf("Expected answer to mention 'dragon', got: %q", ad.Content)
		}
	})

	t.Run("should not emit compaction events when infinite sessions disabled", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		enabled := false
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			InfiniteSessions: &copilot.InfiniteSessionConfig{
				Enabled: &enabled,
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		var compactionEvents []copilot.SessionEvent
		session.On(func(event copilot.SessionEvent) {
			switch event.Data.(type) {
			case *copilot.SessionCompactionStartData, *copilot.SessionCompactionCompleteData:
				compactionEvents = append(compactionEvents, event)
			}
		})

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 2+2?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Should not have any compaction events when disabled
		if len(compactionEvents) != 0 {
			t.Errorf("Expected 0 compaction events when disabled, got %d", len(compactionEvents))
		}
	})
}
